### PR TITLE
Multiline and indentation support inside the parser

### PR DESCRIPTION
**Pull request by ptersilie**
Under development

### CODE REVIEW

**Comment by ltratt** [comment]
Are you looking to merge this or...?

**Comment by ptersilie** [comment]
Not yet. There is still a pesky bug that I'm trying to fix. :(

**Comment by ltratt** [comment]
OK.

**Comment by ptersilie** [comment]
@ltratt Okay, you can have a look now. Once you are done I will squash.

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ save/
 lib/eco/save/
 temp/
 lib/eco/test/php/
+*swp
```
Comment (ltratt) [inline .gitignore]: Should this be `.*swp`?

File: lib/eco/grammar_parser/bootstrap.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 from inclexer.inclexer import IncrementalLexer, IncrementalLexerCF
 from incparser.astree import BOS, EOS
 
-from grammar_parser.gparser import Rule, Nonterminal, Terminal, Epsilon, MagicTerminal
+from grammar_parser.gparser import Rule, Nonterminal, Terminal, Epsilon, MagicTerminal, AnySymbol
```
Comment (ltratt) [inline lib/eco/grammar_parser/bootstrap.py:27]: Wrap long line?

File: lib/eco/incparser/astree.py [hunk 3]
```diff
@@ -275,6 +267,7 @@ def insert_after(self, node):
         self.parent.insert_after_node(self, node)
 
     def insert_after_node(self, node, newnode):
+        # XXX don't mark/save stuff in silent mode
```
Comment (ltratt) [inline lib/eco/incparser/astree.py]: Not sure I understand what this comment is referring to.

File: lib/eco/incparser/incparser.py [hunk 4]
```diff
@@ -253,63 +251,285 @@ def inc_parse(self, line_indents=[], reparse=False):
...
+                    la.indent = list(newindent)
+                    self.last_indent = list(la.indent)
+
+                if self.indents_differ(there, needed):
+                    self.repair_indents(la, there, needed)
+                elif indent_stack_eq:
+                    return
+            self.update_succeeding_lines(la, ws, newindent)
+
+    def update_succeeding_lines(self, la, ws, newindent):
```
Comment (ltratt) [inline lib/eco/incparser/incparser.py:426]: Body of function is double indented?

File: lib/eco/test/test_eco.py [hunk 5]
```diff
@@ -33,6 +33,9 @@
 import pytest
 slow = pytest.mark.slow
 
+import logging
+#logging.getLogger().setLevel(logging.DEBUG)
```
Comment (ltratt) [inline lib/eco/test/test_eco.py]: Delete commented line?

File: .gitignore (hunk 1 above)
Comment (ptersilie) [inline .gitignore]: I probably meant `*.swp`, but I think all of this works. Will change it.

File: lib/eco/test/test_eco.py [hunk 6]
```diff
@@ -2203,3 +2427,26 @@ def test_simple(self):
         assert self.treemanager.export_as_text() == """class X:
     def foo():
         x = SELECT * FROM table"""
+
+class Test_AnySymbol_Indents(Test_Python):
+
```
Comment (ltratt) [inline lib/eco/test/test_eco.py]: Random blank line?

File: lib/eco/incparser/astree.py (hunk 3 above)
Comment (ptersilie) [inline lib/eco/incparser/astree.py]: Whoops. Left over.

**Comment by ltratt** [comment]
OK, I don't have substantive comments, but a few minor things to consider.

File: lib/eco/test/test_eco.py (hunk 5 above)
Comment (ptersilie) [inline lib/eco/test/test_eco.py]: I kind of need to turn this on and off for testing every once in a while I don't always want to lookup the syntax for this. Do you have a better idea to leave it in without using a comment?

Comment (snim2) [inline lib/eco/test/test_eco.py]: In Krun we do this by passing in a command line argument. You can see the code here: https://github.com/softdevteam/krun/blob/master/krun.py (grep for `--debug`)

**Comment by ltratt** [comment]
OK, this looks good to me. Do you want to squash, and then I'll merge?

**Comment by ptersilie** [comment]
Sure. I'll have a look what I can squash.

**Comment by ptersilie** [comment]
I tried to rebase, but would get merge conflicts on every single commit. I'm too scared I'm solving the conflicts wrong which is why I'm just gonna leave things as they are. You can merge if you are happy with this.

**Comment by ltratt** [comment]
I think you need to squash the 'cleanup' merge away (though other things can stay as is). You might want to investigate turning on git rerere.

**Comment by ltratt** [comment]
Oops, I wasn't clear. I mean the cleanup commits in general.

**Comment by ptersilie** [comment]
What is git rerere? I tried squashing them away, but the rebase alone causes ALL commits to have merge conflicts which I wasn't able to solve.

**Comment by ptersilie** [comment]
Turned rerere on, but I still get the same conflicts.

**Comment by ltratt** [comment]
What are you rebasing against?

**Comment by ptersilie** [comment]
master

**Comment by snim2** [comment]
I think you want to do: `git pull --rebase origin master` just to bring this branch up to date with master, then `git rebase -i HEAD~33` to deal with the squashing. Are there a lot of conflicts against the master branch?

**Comment by ptersilie** [comment]
I already merged master into this branch. And then tried `git rebase -i master`. There seem to be quite a few merge conflicts. The main problem is that I have no idea how to fix them, because this code as changed a lot later so I'm not sure which version I need.

And at that point I just gave up, as a few extra commits are not worth risking to break this branch.

**Comment by ltratt** [comment]
I quickly tried a few of these, and the first several merge conflicts are very trivial.

**Comment by snim2** [comment]
If you have already merged master in to this branch (and it looks like you were successful, because GitHub says this PR is merge-able) then you want to rebase the branch to squash your commits, then @ltratt can press the merge button here. So you want:

`git status`

to check you are in branch `multiline_grammar_parserindent`, then:

`git rebase -i HEAD~33`

to squash your commits, then:

`git push -f origin multiline_grammar_parserindent`

to push the changes to the PR. You might want to `cp -a` your repo somewhere first, just in case you have a fall-back if the rebase goes wrong (although you can abort a rebase, but I always feel safer with a copy of the repo around, just in case).

**Comment by ltratt** [comment]
Lukas's problem is a little different. When he does a rebase against master, he gets merge conflicts in various commits along the way. The good news is that -- and I checked the first 10 or so -- they're all really, really simple. So it's annoying to have to resolve these, but not too bad. Once he's done that, he can then squash away the rubbish commits.

As Sarah said, it's perfectly fine to `cp -rp` the repo first if you're nervous. But I think it's important that we do this stuff the right way, so a rebase, followed by squashing is necessary.

**Comment by ptersilie** [comment]
I didn't find them trivial at all. For instance the first one I have no idea which version I should choose.

**Comment by ltratt** [comment]
Why not have a separate copy of the master branch open as well? If in doubt, choose the version that's equivalent to the master branch.

**Comment by ptersilie** [comment]
That's the thing, I did that, but the master branch differs from both options, so it still doesn't help me to know which version master expects me to pick to finally end up at master.

**Comment by ltratt** [comment]
We can't merge unless this is rebased and squashed. I am 100% confident that you can resolve the conflicts correctly; but I am aware that it is a bit of hard work, and can feel a bit scary.

**Comment by ltratt** [comment]
OK, this is definitely a super messy case. I _think_ I may have found a half-way house to make things better without killing ourselves. I'll report back soon.

**Comment by ptersilie** [comment]
Thank you for having a look.

**Comment by ltratt** [comment]
OK, this is definitely too hard to rebase without wasting lots of time, so I'm merging as-is.

**Pull request merged**